autospec/infile_bb_parser.py

```diff
@@ -0,0 +1,234 @@
+        # ??= is the weakest variable assignment, so if that variable already
+        # has an assignment, do not overwrite it. = has the highest precedence
+        # and ?= is between the two.
+        #
+        if key in bb_dict and op == "??=":
+            return bb_dict
+        elif key in bb_dict and op == "?=":
+            if not isinstance(bb_dict[key], list):
+                return bb_dict
+
+        if key in bb_dict and isinstance(bb_dict[key], list):
+            v = bb_dict[key]
+            del [v[1]]
+            bb_dict[key] = "".join(v)
+        try:
+            bb_dict[key] = evaluate_expr(op, bb_dict.get(key), value)
+            if op == "??=":
+                bb_dict[key] = [bb_dict[key], 1]
+        except AttributeError as error:
+            print("Missing operation functionality: {}".format(error))
+
+    return bb_dict
+
+
+def bb_scraper(bb_fp, bb_dict):
+
+    bb_dict["VERSION"] = scrape_version(bb_dict)
+    bb_dict["inherits"] = []
+
+    for line in bb_fp:
+        line = line.strip()
+        if not line or line.startswith('#'):
+            continue
+
+        # If line is a command, create array of command lines
+        # TODO: command could be python code
+        if line.startswith('do_'):
+            cmd_name = line.split('()')[0].strip()
+            cmd = []
+            depth = 0
+            while 1:
+                depth = read_in_command(line, depth)
+                if line:
+                    cmd.append("# " + line)
+
+                if depth == 0:
+                    break
+                else:
+                    line = next(bb_fp).strip()
+
+            # Remove the 'do_command {' and '}' from the list, so that
+            # the dict value is a list of the commands.
+            if cmd[0].startswith("# " + cmd_name) and cmd[-1] == '# }':
+                cmd = cmd[1:-1]
+
+            # Some bitbake files have tasks with prepend or append operations.
+            # Perform the logic to store all tasks with the same 'core' command
+            # name in the dictionary.
+            if not bb_dict.get(cmd_name):
+                bb_dict[cmd_name] = cmd
+            elif cmd_name.rsplit('_') == 'prepend':
+                cmd.extend(bb_dict.get(cmd_name))
+                bb_dict[cmd_name] = cmd
+            else:
+                bb_dict.get(cmd_name).extend(cmd)
+
+            continue
+
+        # If a line continuation, create a string that can easily be split
+        # by '\' for src_uri and other variables.
+        elif line[-1] == '\\':
+            lines = []
+            while 1:
+                lines.append(line)
+                line = next(bb_fp).strip()
+                if line[-1] != '\\':
+                    lines.append(line)
+                    break
+
+            line = " ".join(lines)
+
+        if line.startswith('inherit'):
+            bb_dict['inherits'].append(line.split(maxsplit=1)[1])
+            continue
+
+        match = pattern_match_line(line)
```

Comment:
You are compiling the same regular expression for every line in the file. You can compile that regular expression before this loop and just run search on the line (much faster). Maybe `pattern_match_line()` could return the compiled pattern instead of the result of the regex search?